### PR TITLE
GEODE-9585: adding support for the KEYSLOT command

### PR DESCRIPTION
**Pull request by upthewaterspout**
Adding support for the keyslot command. refactoring the RedisKey code a little
bit to move the slot calculation into a static method that can be used and
tested separately. Changing RedisKey.crc16 to slot, because that is only thing
we are using the crc16 field for.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java [hunk 1]
```diff
@@ -29,41 +28,23 @@
...
-      startHashtag = -1;
-      endHashtag = value.length;
-    }
-
-    crc16 = CRC16.calculate(value, startHashtag + 1, endHashtag);
+    slot = KeyHashUtil.slotForKey(value);
   }
 
   public int getBucketId() {
     // & (REDIS_SLOTS - 1) is equivalent to % REDIS_SLOTS but supposedly faster
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java:46]: remove this comment since it no longer does & (REDIS_SLOTS - 1)

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Utility for parsing the hashtags in a redis key and computing the redis slot of the key.
+ */
+public class KeyHashUtil {
+
+  /**
+   * Compute the slot of a given redis key
+   *
+   * @return the slot, a value between 0 and 16K.
+   */
+  public static int slotForKey(byte[] value) {
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:31]: since your comments refer to a "redis key" should the parameter be named "key" instead of "value"?

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    int startHashtag = Integer.MAX_VALUE;
+    int endHashtag = 0;
+
+    for (int i = 0; i < value.length; i++) {
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:35]: If you have a key that looks like this "aaa}bbb{tag}ccc" this parsing will ignore the "{tag}" and say the whole key is the hash tag. Is that how native redis also does it?

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RegionProvider.java [hunk 4]
```diff
@@ -206,7 +206,7 @@ public RedisData getRedisData(RedisKey key, RedisData notFoundValue) {
 
   private RedisDataMovedException createRedisDataMovedException(RedisKey key) {
     RedisMemberInfo memberInfo = getRedisMemberInfo(key);
-    Integer slot = key.getCrc16() & (REDIS_SLOTS - 1);
+    Integer slot = key.getSlot();
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RegionProvider.java:209]: change "Integer" to "int" here and on RedisDataMovedException

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/parameters/ClusterParameterRequirements.java [hunk 5]
```diff
@@ -27,7 +27,7 @@ public void checkParameters(Command command, ExecutionHandlerContext context) {
 
     if (numberOfArguments < 2) {
       throw new RedisParametersMismatchException(command.wrongNumberOfArgumentsErrorMessage());
-    } else if (numberOfArguments > 2) {
+    } else if (numberOfArguments > 3) {
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/parameters/ClusterParameterRequirements.java:30]: It is not clear to me why this change was made. If it was wrong before should a test have been broken? Can we add a test to the integration test that passes CLUSTER the wrong number of parameters?

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisKeyJUnitTest.java [hunk 6]
```diff
@@ -39,43 +40,45 @@ public static void classSetup() {
...
     key = new RedisKey(new byte[] {});
-    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate(""));
+    assertThat(key.getSlot()).isEqualTo(CRC16.calculate("") % RegionProvider.REDIS_SLOTS);
   }
 
   @Test
   public void testSerialization_withPositiveSignedShortCRC16() throws Exception {
     RedisKey keyOut = new RedisKey(stringToBytes("012345"));
-    assertThat((short) keyOut.getCrc16()).isPositive();
+    assertThat((short) keyOut.getSlot()).isPositive();
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisKeyJUnitTest.java:81]: should this just be "testSerialization" and not worry about the slot being positive as a short? If we wanted to test that slot is always >= 0 it seems like it should be in its own test method.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      if (value[i] == '{' && startHashtag == Integer.MAX_VALUE) {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:36]: In the `StringBytesGlossary` class, there's a comment (originally in the Coder class before the constants were pulled out into their own class) stating:
> Do not use '' <-- java primitive chars. Redis uses `{@link Coder#CHARSET}` encoding so we should
not risk java handling char to byte conversions, rather just hard code `{@link Coder#CHARSET}`
chars as bytes
Is this a concern here?

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java [hunk 8]
```diff
@@ -29,41 +28,23 @@
 import org.apache.geode.internal.serialization.DeserializationContext;
 import org.apache.geode.internal.serialization.KnownVersion;
 import org.apache.geode.internal.serialization.SerializationContext;
-import org.apache.geode.redis.internal.executor.cluster.CRC16;
 import org.apache.geode.redis.internal.executor.cluster.RedisPartitionResolver;
 
 public class RedisKey implements DataSerializableFixedID {
 
-  private int crc16;
+  private int slot;
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java:35]: did you consider changing this field to "short" since its value will always be in the range 0..REDIS_SLOTS-1?

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java [hunk 9]
```diff
@@ -92,15 +73,15 @@ public int getDSFID() {
...
+    out.writeShort(slot);
     DataSerializer.writeByteArray(value, out);
   }
 
   @Override
   public void fromData(DataInput in, DeserializationContext context)
       throws IOException {
     // Need to convert a signed short to unsigned
-    crc16 = in.readShort() & 0xffff;
+    slot = in.readShort() & 0xffff;
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java:84]: since slot will never be negative I think you can drop the "& 0xffff".

**Review by DonalEvans — APPROVED** [review]

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+package org.apache.geode.redis.internal.data;
+
+import static org.apache.geode.redis.internal.RegionProvider.REDIS_SLOTS;
+
+import org.apache.geode.redis.internal.executor.cluster.CRC16;
+
+/**
+ * Utility for parsing the hashtags in a redis key and computing the redis slot of the key.
+ */
+public class KeyHashUtil {
```
Comment (kirklund) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:24]: I think a unit test for `KeyHashUtil` would be valuable and pretty easy to write.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java (hunk 3 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:35]: Oooh, good catch! BTW - I didn't write this hashing logic, I just moved it to a new class. But I added a test and it looks like native redis handles this differently!

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/parameters/ClusterParameterRequirements.java (hunk 5 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/parameters/ClusterParameterRequirements.java:30]: I've added some more tests to the `AbstractClusterIntegrationTest`. I changed this `ClusterParameterRequirements` class because the command `CLUSTER KEYSLOT XXX` has three arguments. I added additional validation into the `ClusterExecutor`

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java (hunk 1 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java (hunk 2 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java (hunk 3 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java]: I fixed this. I changed the hash implementation to be closer to native redis.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RegionProvider.java (hunk 4 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RegionProvider.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java (hunk 8 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java (hunk 9 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java (hunk 10 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java:26]: I added a new test. This code was previously covered by the RedisKey unit test. I moved that testing into KeyHashUtilTest and added more coverage.

File: geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisKeyJUnitTest.java (hunk 6 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/data/RedisKeyJUnitTest.java]: done

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java (hunk 7 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java]: I put a couple of constants in StringBytesGlossary.

However, I don't think it is necessary for that class to hard code the decimal value of these chars ( eg 56, 48, etc.). `char` literals in java are always UTF-16. So converting them to a byte will always have the same value. That is not true for String.getBytes(), which uses a OS platform specific encoding.

So what I changed is that the constants are in still StringBytesGlossary but I changed that class to use char literals, which I think makes it easier to read and maintain.

Comment (DonalEvans) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java]: Ah, thanks for the explanation. Yeah, that definitely makes the class more readable. If it's not too much trouble, there are some other single-char constants in that file starting on line 243 that could also be simplified in this way.

Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/KeyHashUtil.java]: Done!

**Comment by upthewaterspout** [comment]
@ezoerner - added the randomized integration test.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/netty/StringBytesGlossary.java [hunk 11]
```diff
@@ -32,27 +32,37 @@
   /**
    * byte identifier of a bulk string
    */
-  public static final byte BULK_STRING_ID = 36; // '$'
+  public static final byte BULK_STRING_ID = '$';
```
Comment (dschneider-pivotal) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/netty/StringBytesGlossary.java:35]: Is a java character literal's ordinal value stable? Or can it change based on the locale? If you know the answer can you send a link?

**Review by dschneider-pivotal — APPROVED** [review]

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/netty/StringBytesGlossary.java (hunk 11 above)
Comment (upthewaterspout) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/netty/StringBytesGlossary.java:35]: The only link I have is the java language spec - https://docs.oracle.com/javase/specs/jls/se16/html/jls-3.html#jls-3.10.4. The character literals are UTF-16.
